drivers: usb: udc: numaker: fix some issues

ccli8 opened the pull request:
This fixes some issues of nuvoton numaker udc driver, including:
1. Fix timing to enable D+ pull-up so that device won't be recognized by host until user application invokes `udc_enable()`.
1. Refine interrupt handling sequence: clear interrupt flag first, then process it.
   This can avoid one race condition when interrupt flag is cleared but not processed for interrupts of the same type coming very closely.
1. Support VBUS detect (`caps.can_detect_vbus`). This requires enabling VBUS detect interrupt early.

ccli8 committed:
drivers: usb: udc: numaker: fix typo

This fixes typo in isr function name.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: fix timing to enable D+ pull-up

This fixes timing to enable D+ pull-up so that device won't be recognized
by host until user application invokes udc_enable().

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: support VBUS detect

This supports VBUS detect (caps.can_detect_vbus). This requires enables
VBUS detect interrupt early.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: refine timing to clear interrupt flags

This refines interrupt handling sequence: clear interrupt flag first,
then process it. This can avoid one race condition when interrupt flag
is cleared but not processed for interrupts of the same type coming
very closely.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -172,8 +172,7 @@ static inline void numaker_usbd_sw_connect(const struct device *dev)
 	base->INTEN = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET |
-		      IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USBD_INT_SOF |))
-		      USBD_INT_WAKEUP;
+		      IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USBD_INT_SOF |)) USBD_INT_WAKEUP;
```
The change is caused by **clang-format** tool. I move out the `CONFIG_UDC_ENABLE_SOF` code to separate `#if` to make more readable.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -1216,12 +1216,22 @@ static void numaker_usbd_isr(const struct device *dev)
 
 	uint32_t volatile usbd_intsts = base->INTSTS;
 	uint32_t volatile usbd_bus_state = base->ATTR;
+	uint32_t usbd_inten_all =
```
I see 'volatile' above, please remove them in a separate commit.

ccli8 replied:
Removed `volatile`

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -1216,12 +1216,22 @@ static void numaker_usbd_isr(const struct device *dev)
 
 	uint32_t volatile usbd_intsts = base->INTSTS;
 	uint32_t volatile usbd_bus_state = base->ATTR;
+	uint32_t usbd_inten_all =
+		USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET | USBD_INT_SOF | USBD_INT_WAKEUP;
```
Using bad tools is no excuse for not formatting nicely.
```c
	const uint32_t usbd_inten_all = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET |
					USBD_INT_SOF | USBD_INT_WAKEUP;
```

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -1215,12 +1215,18 @@ static void numaker_usbd_isr(const struct device *dev)
+		USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET | USBD_INT_SOF | USBD_INT_WAKEUP;
+
+	/* Focus on enabled */
+	usbd_intsts =
+		(usbd_intsts & ~usbd_inten_all) | (usbd_intsts & base->INTEN & usbd_inten_all);
```
It does not look right. You can store the flags of the handled interrupts in a variable and use that variable to clear base->INTSTS at the end of the ISR handling function.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -171,9 +171,10 @@ static inline void numaker_usbd_sw_connect(const struct device *dev)
-	base->INTEN = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET |
-		      IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USBD_INT_SOF |))
-		      USBD_INT_WAKEUP;
+	base->INTEN = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET | USBD_INT_WAKEUP;
+#if defined(CONFIG_UDC_ENABLE_SOF)
+	base->INTEN |= USBD_INT_SOF;
+#endif
```
>This refines code style on CONFIG_UDC_ENABLE_SOF to be more readable.

No, it absolutely does not. Please keep it as it is.

ccli8 replied:
Reverted back this change

ccli8 committed:
drivers: usb: udc: numaker: recover unwanted assignment break

This recovers unwanted assignment break caused by clang-format.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: remove unneeded volatile

This removes unneeded 'volatile' qualifier for normal memory variables.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -345,8 +346,8 @@ static int numaker_usbd_hw_setup(const struct device *dev)
-	base->ATTR = USBD_ATTR_BYTEM_Msk | BIT(9) | USBD_ATTR_DPPUEN_Msk | USBD_ATTR_USBEN_Msk |
-		     BIT(6) | USBD_ATTR_PHYEN_Msk;
+	base->ATTR =
+		USBD_ATTR_BYTEM_Msk | BIT(9) | USBD_ATTR_USBEN_Msk | BIT(6) | USBD_ATTR_PHYEN_Msk;
```
There is logic change here. D+ pull-up enable is moved out to `udc_enable` so that host won't recognize device until `udc_enable` is invoked.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -172,7 +172,7 @@ static inline void numaker_usbd_sw_connect(const struct device *dev)
 	/* Enable relevant interrupts */
 	base->INTEN = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET |
-		      IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USBD_INT_SOF |))
+		      IF_ENABLED(CONFIG_UDC_ENABLE_SOF, (USBD_INT_SOF |)) /* High CPU load */
```
"High CPU load" comment is a bit extreme. It depends on the bus speed, the CPU, and the tasks that need to be performed during the SOF event. However, that is not relevant here.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -1371,8 +1371,8 @@ static void numaker_udbd_isr(const struct device *dev)
 				struct numaker_usbd_ep *ep_ctrlout = priv->ep_pool + 0;
-				USBD_EP_T *ep_ctrlout_base =
-					numaker_usbd_ep_base(dev, ep_ctrlout->ep_hw_idx);
+				USBD_EP_T *ep_ctrlout_base = numaker_usbd_ep_base(
+					dev, ep_ctrlout->ep_hw_idx);
```
This kind of reformatting does not make it better. numaker_udbd_isr() function is quite large, I would split it in a few smaller and more specific routines, plug/unplug, wakeup, setup and transaction/transfer handling. Better code has fewer indents.

ccli8 committed:
drivers: usb: udc: numaker: refactor large isr function

For clear, this splits large isr function into smaller ones which are
dedicated for handling plug/unplug, wakeup, reset/suspend/resume, SOF,
Setup, and endpoint events.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented:
@jfischer-no Please review if you are available

ccli8 commented:
More comment?

jfischer-no reviewed (approved):
LGTM
